File: DataStructures/Trees/BinaryIndexedTree/FenwickTree/atcoder/A59/A59.php
<?php

class FenwickTree
{
    private int $n;
    private array $tree;

    public function __construct(int $n)
    {
        $this->n = $n;
        $this->tree = array_fill(0, $n + 2, 0);
    }

    public function add(int $index, int $value): void
    {
        while ($index <= $this->n) {
            $this->tree[$index] += $value;
            $index += $index & -$index;
        }
    }

    public function sum(int $index): int
    {
        $res = 0;
        while ($index > 0) {
            $res += $this->tree[$index];
            $index -= $index & -$index;
        }
        return $res;
    }

    public function rangeSum(int $l, int $r): int
    {
        return $this->sum($r - 1) - $this->sum($l - 1);
    }
}

$input = stream_get_contents(STDIN);
$tokens = explode("\n", trim($input));
list($N, $Q) = array_map('intval', explode(' ', array_shift($tokens)));

$bit = new FenwickTree($N);
$A = array_fill(0, $N + 1, 0); // 1-indexed
$output = [];

foreach ($tokens as $line) {
    if (trim($line) === '') continue; // 空行スキップ

    $parts = explode(' ', $line);
    $c = intval($parts[0]);

    if ($c === 1) {
        $pos = intval($parts[1]);
        $x = intval($parts[2]);
        $diff = $x - $A[$pos];
        $A[$pos] = $x;
        $bit->add($pos, $diff);
    } elseif ($c === 2) {
        $l = intval($parts[1]);
        $r = intval($parts[2]);
        $output[] = $bit->rangeSum($l, $r);
    }
}

echo implode("\n", $output) . "\n";

File: DataStructures/Trees/BinaryIndexedTree/FenwickTree/atcoder/A59/A59.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type FenwickTree struct {
	n    int
	tree []int
}

func NewFenwickTree(size int) *FenwickTree {
	return &FenwickTree{
		n:    size,
		tree: make([]int, size+1), // 1-indexed
	}
}

func (f *FenwickTree) Add(index, value int) {
	for index <= f.n {
		f.tree[index] += value
		index += index & -index
	}
}

func (f *FenwickTree) Sum(index int) int {
	res := 0
	for index > 0 {
		res += f.tree[index]
		index -= index & -index
	}
	return res
}

func (f *FenwickTree) RangeSum(l, r int) int {
	return f.Sum(r-1) - f.Sum(l-1)
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	scanner := bufio.NewScanner(reader)
	scanner.Split(bufio.ScanLines)

	// 最初の行: N Q
	scanner.Scan()
	parts := strings.Fields(scanner.Text())
	N, _ := strconv.Atoi(parts[0])
	Q, _ := strconv.Atoi(parts[1])

	ft := NewFenwickTree(N)
	A := make([]int, N+1) // 1-indexed

	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	for i := 0; i < Q && scanner.Scan(); i++ {
		line := scanner.Text()
		query := strings.Fields(line)

		if query[0] == "1" {
			pos, _ := strconv.Atoi(query[1])
			x, _ := strconv.Atoi(query[2])
			diff := x - A[pos]
			A[pos] = x
			ft.Add(pos, diff)
		} else if query[0] == "2" {
			l, _ := strconv.Atoi(query[1])
			r, _ := strconv.Atoi(query[2])
			sum := ft.RangeSum(l, r)
			fmt.Fprintln(writer, sum)
		}
	}
}
